File: clients/python/src/beebium/keyboard.py
# Copyright 2025 Robert Smallshire <robert@smallshire.org.uk>
#
# This file is part of Beebium.
#
# Beebium is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version. Beebium is distributed in the hope that it will
# be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with Beebium.
# If not, see <https://www.gnu.org/licenses/>.

"""Keyboard input interface for the beebium client."""

from __future__ import annotations

import time
from dataclasses import dataclass

from beebium._proto import keyboard_pb2, keyboard_pb2_grpc
from beebium.keyboard_map import (
    CTRL_KEY,
    DELETE_KEY,
    ESCAPE_KEY,
    RETURN_KEY,
    SHIFT_KEY,
    SPACE_KEY,
    char_to_matrix,
)


@dataclass
class KeyboardState:
    """Current state of the keyboard matrix."""

    pressed_rows: list[int]  # 10 rows of pressed key bitmaps

    def is_pressed(self, row: int, column: int) -> bool:
        """Check if a specific key is pressed.

        Args:
            row: The keyboard matrix row (0-9).
            column: The keyboard matrix column (0-9).

        Returns:
            True if the key is pressed.
        """
        if 0 <= row < len(self.pressed_rows):
            return bool(self.pressed_rows[row] & (1 << column))
        return False


class Keyboard:
    """Keyboard input interface.

    Provides both low-level matrix access and high-level text input.

    Usage:
        # Type text
        bbc.keyboard.type("PRINT 42")
        bbc.keyboard.press_return()

        # Press specific keys
        bbc.keyboard.key_down('A')
        bbc.keyboard.key_up('A')

        # Matrix-level access
        bbc.keyboard.matrix_down(row=4, column=1)  # 'A' key
    """

    def __init__(self, stub: keyboard_pb2_grpc.KeyboardServiceStub):
        """Create a keyboard interface.

        Args:
            stub: The gRPC stub for the KeyboardService.
        """
        self._stub = stub
        self._pressed_keys: set[tuple[str, bool]] = set()

    # High-level text input

    def type(
        self,
        text: str,
        inter_key_delay: float = 0.05,
        release_delay: float = 0.02,
    ) -> None:
        """Type a string of text.

        Each character is pressed and released with configurable delays.
        Handles shift automatically for uppercase and symbols.

        Args:
            text: The text to type.
            inter_key_delay: Delay between key presses (seconds).
            release_delay: Delay between press and release (seconds).
        """
        for char in text:
            if self.key_down(char):
                time.sleep(release_delay)
                self.key_up(char)
                time.sleep(inter_key_delay)

    def press_return(self, hold_time: float = 0.02) -> None:
        """Press and release the RETURN key.

        Args:
            hold_time: How long to hold the key (seconds).
        """
        self.matrix_down(*RETURN_KEY)
        time.sleep(hold_time)
        self.matrix_up(*RETURN_KEY)

    def press_escape(self, hold_time: float = 0.02) -> None:
        """Press and release the ESCAPE key.

        Args:
            hold_time: How long to hold the key (seconds).
        """
        self.matrix_down(*ESCAPE_KEY)
        time.sleep(hold_time)
        self.matrix_up(*ESCAPE_KEY)

    def press_delete(self, hold_time: float = 0.02) -> None:
        """Press and release the DELETE key.

        Args:
            hold_time: How long to hold the key (seconds).
        """
        self.matrix_down(*DELETE_KEY)
        time.sleep(hold_time)
        self.matrix_up(*DELETE_KEY)

    def press_space(self, hold_time: float = 0.02) -> None:
        """Press and release the SPACE key.

        Args:
            hold_time: How long to hold the key (seconds).
        """
        self.matrix_down(*SPACE_KEY)
        time.sleep(hold_time)
        self.matrix_up(*SPACE_KEY)

    # Character-level input

    def key_down(self, char: str) -> bool:
        """Press a key for the given character.

        Automatically handles SHIFT for uppercase and shifted symbols.

        Args:
            char: The character to press.

        Returns:
            True if the character is mapped, False otherwise.
        """
        mapping = char_to_matrix(char)
        if mapping is None:
            return False

        row, column, needs_shift = mapping

        if needs_shift:
            self.matrix_down(*SHIFT_KEY)
        self.matrix_down(row, column)

        self._pressed_keys.add((char, needs_shift))
        return True

    def key_up(self, char: str) -> bool:
        """Release a key for the given character.

        Args:
            char: The character to release.

        Returns:
            True if the character is mapped, False otherwise.
        """
        mapping = char_to_matrix(char)
        if mapping is None:
            return False

        row, column, needs_shift = mapping

        self.matrix_up(row, column)

        # Only release shift if no other shifted keys are pressed
        if needs_shift:
            self._pressed_keys.discard((char, True))
            if not any(shifted for _, shifted in self._pressed_keys):
                self.matrix_up(*SHIFT_KEY)
        else:
            self._pressed_keys.discard((char, False))

        return True

    def release_all(self) -> None:
        """Release all currently pressed keys."""
        for char, _ in list(self._pressed_keys):
            self.key_up(char)

    # Modifier keys

    def shift_down(self) -> None:
        """Press the SHIFT key."""
        self.matrix_down(*SHIFT_KEY)

    def shift_up(self) -> None:
        """Release the SHIFT key."""
        self.matrix_up(*SHIFT_KEY)

    def ctrl_down(self) -> None:
        """Press the CTRL key."""
        self.matrix_down(*CTRL_KEY)

    def ctrl_up(self) -> None:
        """Release the CTRL key."""
        self.matrix_up(*CTRL_KEY)

    # Matrix-level input

    def matrix_down(self, row: int, column: int) -> bool:
        """Press a key by BBC keyboard matrix position.

        Args:
            row: The keyboard matrix row (0-9).
            column: The keyboard matrix column (0-9).

        Returns:
            True if accepted by server.
        """
        request = keyboard_pb2.KeyRequest(ik_number=(row << 4) | column)
        response = self._stub.KeyDown(request)
        return response.accepted

    def matrix_up(self, row: int, column: int) -> bool:
        """Release a key by BBC keyboard matrix position.

        Args:
            row: The keyboard matrix row (0-9).
            column: The keyboard matrix column (0-9).

        Returns:
            True if accepted by server.
        """
        request = keyboard_pb2.KeyRequest(ik_number=(row << 4) | column)
        response = self._stub.KeyUp(request)
        return response.accepted

    def get_state(self) -> KeyboardState:
        """Get current keyboard state (pressed keys bitmap).

        Returns:
            The current keyboard state.
        """
        request = keyboard_pb2.GetStateRequest()
        response = self._stub.GetState(request)
        return KeyboardState(pressed_rows=list(response.pressed_rows))

    # Type-ahead (TypeQuickly)

    def type_quickly(self, text: str, cycles_per_key: int = 0) -> int:
        """Enqueue text for typing at machine speed (non-blocking).

        The text is added to a queue and typed character-by-character
        as the emulator runs. Returns immediately.

        Args:
            text: The text to type.
            cycles_per_key: CPU cycles per keystroke (0 = use default 100000).

        Returns:
            Total pending characters in queue after enqueue.

        Raises:
            ValueError: If text contains unmappable characters.
        """
        request = keyboard_pb2.TypeQuicklyRequest(
            text=text,
            cycles_per_key=cycles_per_key,
        )
        response = self._stub.TypeQuickly(request)
        if not response.accepted:
            raise ValueError(f"Text rejected: {response.error}")
        return response.pending_characters

    def typing_status(self) -> tuple[bool, int, int]:
        """Get type-ahead queue status.

        Returns:
            Tuple of (idle, pending_characters, strings_queued).
        """
        request = keyboard_pb2.GetTypingStatusRequest()
        response = self._stub.GetTypingStatus(request)
        return (response.idle, response.pending_characters, response.strings_queued)

    def clear_typing(self) -> int:
        """Clear the type-ahead queue.

        Returns:
            Number of characters that were cleared.
        """
        request = keyboard_pb2.ClearTypingRequest()
        response = self._stub.ClearTyping(request)
        return response.characters_cleared

    def wait_for_typing(self, poll_interval: float = 0.01) -> None:
        """Block until type-ahead queue is empty.

        Args:
            poll_interval: Seconds between status checks.
        """
        while True:
            idle, _, _ = self.typing_status()
            if idle:
                return
            time.sleep(poll_interval)

    # Character-to-key mapping (server-side canonical mapping)

    def get_key_mapping(self, char: str) -> tuple[int, bool, str] | None:
        """Get the BBC key mapping for a character from the server.

        This queries the canonical mapping table maintained by the server.

        Args:
            char: A single character.

        Returns:
            Tuple of (ik_number, needs_shift, name), or None if unmappable.
        """
        request = keyboard_pb2.GetKeyMappingRequest(character=char)
        entry = self._stub.GetKeyMapping(request)
        if not entry.found:
            return None
        return (entry.ik_number, entry.needs_shift, entry.name)

    def get_all_mappings(self) -> list[tuple[str, int, bool, str]]:
        """Get the complete key mapping table from the server.

        Returns:
            List of (character, ik_number, needs_shift, name) tuples.
            Character is empty string for named-only keys (function keys, etc.).
        """
        request = keyboard_pb2.GetAllKeyMappingsRequest()
        response = self._stub.GetAllKeyMappings(request)
        return [
            (entry.character, entry.ik_number, entry.needs_shift, entry.name)
            for entry in response.mappings
        ]

    def is_typeable_on_server(self, text: str) -> bool:
        """Check if all characters in text can be typed (using server mapping).

        Args:
            text: The text to check.

        Returns:
            True if all characters are typeable according to server.
        """
        return all(self.get_key_mapping(c) is not None for c in text)

File: clients/typescript/src/keyboard.ts
/**
 * Keyboard input interface for the Beebium client.
 */

import { promisify } from "./call-utils.js";
import {
    CTRL_KEY,
    DELETE_KEY,
    ESCAPE_KEY,
    RETURN_KEY,
    SHIFT_KEY,
    SPACE_KEY,
    charToMatrix,
} from "./keyboard-map.js";
import type {
    AllKeyMappingsResponse,
    BreakDownResponse,
    BreakKeyState,
    BreakUpResponse,
    ClearTypingResponse,
    KeyMappingEntry,
    KeyResponse,
    KeyboardServiceClient,
    KeyboardState as KeyboardStateProto,
    LinksState,
    SetLinksResponse,
    SetStartupAutoBootResponse,
    SetStartupScreenModeResponse,
    StartupAutoBootState,
    StartupScreenModeState,
    TypeQuicklyResponse,
    TypingStatus,
} from "./generated/keyboard.js";

export {
    CTRL_KEY,
    DELETE_KEY,
    ESCAPE_KEY,
    RETURN_KEY,
    SHIFT_KEY,
    SPACE_KEY,
    charToMatrix,
    matrixToChar,
} from "./keyboard-map.js";

/**
 * Current state of the keyboard matrix.
 */
export interface KeyboardState {
    /** Bitmap of pressed keys (10 rows). */
    pressedRows: number[];

    /** Check if a specific key is pressed. */
    isPressed(row: number, column: number): boolean;
}

function createKeyboardState(pressedRows: number[]): KeyboardState {
    return {
        pressedRows,
        isPressed(row: number, column: number): boolean {
            if (row >= 0 && row < pressedRows.length) {
                const rowValue = pressedRows[row];
                return rowValue !== undefined && (rowValue & (1 << column)) !== 0;
            }
            return false;
        },
    };
}

/**
 * Keyboard input interface.
 *
 * Provides both low-level matrix access and high-level text input.
 *
 * Usage:
 *     // Type text (cycle-paced, use \r for RETURN)
 *     await keyboard.type("PRINT 42\r");
 *
 *     // Press specific keys
 *     await keyboard.keyDown('A');
 *     await keyboard.keyUp('A');
 *
 *     // Matrix-level access
 *     await keyboard.matrixDown(4, 1);  // 'A' key
 */
export class Keyboard {
    private readonly _stub: KeyboardServiceClient;
    private readonly _pressedKeys = new Set<string>();

    constructor(stub: KeyboardServiceClient) {
        this._stub = stub;
    }

    // =========================================================================
    // High-level text input (cycle-paced via server-side TypeAheadQueue)
    // =========================================================================

    /**
     * Type a string of text, paced in emulator cycles.
     *
     * The text is enqueued and typed character-by-character as the
     * emulator runs. Use `\r` for RETURN, `\x1b` for ESCAPE,
     * `\x7f` for DELETE, and `\t` for TAB.
     *
     * Returns immediately; the emulator must be running for the
     * keystrokes to be processed.
     *
     * @param text - The text to type.
     * @param cyclesPerKey - CPU cycles per keystroke (0 = use server default).
     * @returns Total pending characters in queue after enqueue.
     * @throws Error if text contains unmappable characters.
     */
    async type(text: string, cyclesPerKey: number = 0): Promise<number> {
        const response = await promisify<
            { text: string; cyclesPerKey: number },
            TypeQuicklyResponse
        >(this._stub as unknown as Record<string, Function>, "typeQuickly", {
            text,
            cyclesPerKey,
        });
        if (!response.accepted) {
            throw new Error(`Text rejected: ${response.error}`);
        }
        return response.pendingCharacters;
    }

    /** Enqueue a RETURN keypress. */
    async pressReturn(): Promise<void> {
        await this.type("\r");
    }

    /** Enqueue an ESCAPE keypress. */
    async pressEscape(): Promise<void> {
        await this.type("\x1b");
    }

    /** Enqueue a DELETE keypress. */
    async pressDelete(): Promise<void> {
        await this.type("\x7f");
    }

    /** Enqueue a SPACE keypress. */
    async pressSpace(): Promise<void> {
        await this.type(" ");
    }

    // =========================================================================
    // Character-level input
    // =========================================================================

    /**
     * Press a key for the given character.
     *
     * Automatically handles SHIFT for uppercase and shifted symbols.
     *
     * @param char - The character to press.
     * @returns True if the character is mapped, false otherwise.
     */
    async keyDown(char: string): Promise<boolean> {
        const mapping = charToMatrix(char);
        if (mapping === undefined) {
            return false;
        }

        const [row, column, needsShift] = mapping;

        if (needsShift) {
            await this.matrixDown(...SHIFT_KEY);
        }
        await this.matrixDown(row, column);

        this._pressedKeys.add(`${char}:${needsShift}`);
        return true;
    }

    /**
     * Release a key for the given character.
     *
     * @param char - The character to release.
     * @returns True if the character is mapped, false otherwise.
     */
    async keyUp(char: string): Promise<boolean> {
        const mapping = charToMatrix(char);
        if (mapping === undefined) {
            return false;
        }

        const [row, column, needsShift] = mapping;

        await this.matrixUp(row, column);

        // Only release shift if no other shifted keys are pressed
        if (needsShift) {
            this._pressedKeys.delete(`${char}:true`);
            let anyShifted = false;
            for (const key of this._pressedKeys) {
                if (key.endsWith(":true")) {
                    anyShifted = true;
                    break;
                }
            }
            if (!anyShifted) {
                await this.matrixUp(...SHIFT_KEY);
            }
        } else {
            this._pressedKeys.delete(`${char}:false`);
        }

        return true;
    }

    /** Release all currently pressed keys. */
    async releaseAll(): Promise<void> {
        const keys = [...this._pressedKeys];
        for (const key of keys) {
            const char = key.substring(0, key.lastIndexOf(":"));
            await this.keyUp(char);
        }
    }

    // =========================================================================
    // Modifier keys
    // =========================================================================

    /** Press the SHIFT key. */
    async shiftDown(): Promise<void> {
        await this.matrixDown(...SHIFT_KEY);
    }

    /** Release the SHIFT key. */
    async shiftUp(): Promise<void> {
        await this.matrixUp(...SHIFT_KEY);
    }

    /** Press the CTRL key. */
    async ctrlDown(): Promise<void> {
        await this.matrixDown(...CTRL_KEY);
    }

    /** Release the CTRL key. */
    async ctrlUp(): Promise<void> {
        await this.matrixUp(...CTRL_KEY);
    }

    // =========================================================================
    // Matrix-level input
    // =========================================================================

    /**
     * Press a key by BBC keyboard matrix position.
     *
     * @param row - The keyboard matrix row (0-7).
     * @param column - The keyboard matrix column (0-9).
     * @returns True if accepted by server.
     */
    async matrixDown(row: number, column: number): Promise<boolean> {
        const response = await promisify<{ ikNumber: number }, KeyResponse>(
            this._stub as unknown as Record<string, Function>,
            "keyDown",
            { ikNumber: (row << 4) | column },
        );
        return response.accepted;
    }

    /**
     * Release a key by BBC keyboard matrix position.
     *
     * @param row - The keyboard matrix row (0-7).
     * @param column - The keyboard matrix column (0-9).
     * @returns True if accepted by server.
     */
    async matrixUp(row: number, column: number): Promise<boolean> {
        const response = await promisify<{ ikNumber: number }, KeyResponse>(
            this._stub as unknown as Record<string, Function>,
            "keyUp",
            { ikNumber: (row << 4) | column },
        );
        return response.accepted;
    }

    /**
     * Get current keyboard state (pressed keys bitmap).
     *
     * @returns The current keyboard state.
     */
    async getState(): Promise<KeyboardState> {
        const response = await promisify<Record<string, never>, KeyboardStateProto>(
            this._stub as unknown as Record<string, Function>,
            "getState",
            {},
        );
        return createKeyboardState([...response.pressedRows]);
    }

    // =========================================================================
    // Type-ahead status
    // =========================================================================

    /**
     * Get type-ahead queue status.
     *
     * @returns Object with idle, pendingCharacters, and stringsQueued.
     */
    async typingStatus(): Promise<{ idle: boolean; pendingCharacters: number; stringsQueued: number }> {
        const response = await promisify<Record<string, never>, TypingStatus>(
            this._stub as unknown as Record<string, Function>,
            "getTypingStatus",
            {},
        );
        return {
            idle: response.idle,
            pendingCharacters: response.pendingCharacters,
            stringsQueued: response.stringsQueued,
        };
    }

    /**
     * Clear the type-ahead queue.
     *
     * @returns Number of characters that were cleared.
     */
    async clearTyping(): Promise<number> {
        const response = await promisify<Record<string, never>, ClearTypingResponse>(
            this._stub as unknown as Record<string, Function>,
            "clearTyping",
            {},
        );
        return response.charactersCleared;
    }

    /**
     * Wait until the type-ahead queue is empty.
     *
     * @param pollIntervalMs - Milliseconds between status checks (default 10).
     */
    async waitForTyping(pollIntervalMs: number = 10): Promise<void> {
        while (true) {
            const status = await this.typingStatus();
            if (status.idle) {
                return;
            }
            await new Promise(r => setTimeout(r, pollIntervalMs));
        }
    }

    // =========================================================================
    // Character-to-key mapping (server-side canonical mapping)
    // =========================================================================

    /**
     * Get the BBC key mapping for a character from the server.
     *
     * This queries the canonical mapping table maintained by the server.
     *
     * @param char - A single character.
     * @returns Object with ikNumber, needsShift, and name, or undefined if unmappable.
     */
    async getKeyMapping(char: string): Promise<{ ikNumber: number; needsShift: boolean; name: string } | undefined> {
        const response = await promisify<{ character: string }, KeyMappingEntry>(
            this._stub as unknown as Record<string, Function>,
            "getKeyMapping",
            { character: char },
        );
        if (!response.found) {
            return undefined;
        }
        return {
            ikNumber: response.ikNumber,
            needsShift: response.needsShift,
            name: response.name,
        };
    }

    /**
     * Get the complete key mapping table from the server.
     *
     * @returns Array of mapping entries with character, ikNumber, needsShift, and name.
     *          Character is empty string for named-only keys (function keys, etc.).
     */
    async getAllMappings(): Promise<Array<{ character: string; ikNumber: number; needsShift: boolean; name: string }>> {
        const response = await promisify<Record<string, never>, AllKeyMappingsResponse>(
            this._stub as unknown as Record<string, Function>,
            "getAllKeyMappings",
            {},
        );
        return response.mappings.map(entry => ({
            character: entry.character,
            ikNumber: entry.ikNumber,
            needsShift: entry.needsShift,
            name: entry.name,
        }));
    }

    /**
     * Check if all characters in text can be typed (using server mapping).
     *
     * @param text - The text to check.
     * @returns True if all characters are typeable according to the server.
     */
    async isTypeableOnServer(text: string): Promise<boolean> {
        for (const char of text) {
            const mapping = await this.getKeyMapping(char);
            if (mapping === undefined) {
                return false;
            }
        }
        return true;
    }

    // =========================================================================
    // Break key operations
    // =========================================================================
    // The Break key is NOT part of the keyboard matrix. It is directly
    // connected to the reset circuit (IC16 NE555 timer) via pin 4.
    //
    // While Break is held: CPU is halted (reset line held low)
    // On Break release: Soft reset sequence begins
    //
    // Note: Hardware always does a soft reset. MOS checks if Ctrl is held
    // during its reset sequence to decide between warm/cold reset behavior.

    /**
     * Hold the Break key (halt the CPU).
     *
     * This asserts the reset line, halting the CPU. The CPU remains
     * halted until breakUp() is called.
     *
     * @returns True if successful.
     */
    async breakDown(): Promise<boolean> {
        const response = await promisify<Record<string, never>, BreakDownResponse>(
            this._stub as unknown as Record<string, Function>,
            "breakDown",
            {},
        );
        return response.success;
    }

    /**
     * Release the Break key (begin soft reset sequence).
     *
     * This releases the reset line. The CPU will execute the 7-cycle
     * reset sequence and then begin execution from the reset vector.
     *
     * MOS checks if Ctrl is held at this moment to determine reset type:
     * - If Ctrl pressed: MOS performs "hard reset" (clears VIA config)
     * - If Ctrl not pressed: MOS performs "warm reset" (preserves state)
     *
     * @returns True if successful.
     */
    async breakUp(): Promise<boolean> {
        const response = await promisify<Record<string, never>, BreakUpResponse>(
            this._stub as unknown as Record<string, Function>,
            "breakUp",
            {},
        );
        return response.success;
    }

    /**
     * Check if the Break key is currently held.
     *
     * @returns True if Break is currently held (CPU halted).
     */
    async isBreakHeld(): Promise<boolean> {
        const response = await promisify<Record<string, never>, BreakKeyState>(
            this._stub as unknown as Record<string, Function>,
            "getBreakState",
            {},
        );
        return response.isHeld;
    }

    /**
     * Press and release the Break key (perform soft reset).
     *
     * This is a convenience method that calls breakDown(), waits
     * briefly, then calls breakUp().
     *
     * @param holdTimeMs - How long to hold Break in milliseconds (default 20).
     * @returns True if both operations succeeded.
     */
    async pressBreak(holdTimeMs: number = 20): Promise<boolean> {
        const downOk = await this.breakDown();
        await new Promise(r => setTimeout(r, holdTimeMs));
        const upOk = await this.breakUp();
        return downOk && upOk;
    }

    /**
     * Press Ctrl-Break (trigger MOS hard reset).
     *
     * This holds Ctrl while pressing Break, which causes MOS to
     * detect a "hard reset" and perform full reinitialization.
     *
     * Note: The hardware always performs a soft reset. MOS checks
     * the keyboard matrix during its reset routine and clears the
     * VIA configuration if Ctrl is held, simulating a hard reset.
     *
     * @param holdTimeMs - How long to hold Break in milliseconds (default 20).
     * @returns True if all operations succeeded.
     */
    async ctrlBreak(holdTimeMs: number = 20): Promise<boolean> {
        await this.ctrlDown();
        await new Promise(r => setTimeout(r, 10)); // Brief delay to ensure Ctrl is registered
        const downOk = await this.breakDown();
        await new Promise(r => setTimeout(r, holdTimeMs));
        const upOk = await this.breakUp();
        await new Promise(r => setTimeout(r, 10)); // Brief delay before releasing Ctrl
        await this.ctrlUp();
        return downOk && upOk;
    }

    // =========================================================================
    // Keyboard links (DIP switches)
    // =========================================================================
    // The BBC Micro has 8 keyboard links (row 0, columns 2-9) that form a
    // startup options byte. Active-low: bit SET = link broken (open),
    // bit CLEAR = link made.
    //
    // Bit layout:
    //   Bits 0-2: Screen mode (XOR'd with 7 by MOS, so 0x07 = Mode 7)
    //   Bit 3: SHIFT-BREAK action (1 = normal, 0 = reversed/auto-boot)
    //   Bits 4-7: ROM-dependent (disc timing, filing system, etc.)
    //
    // Default value 0xFF (all links broken) = Mode 7, normal SHIFT-BREAK.

    /**
     * Get the raw keyboard links byte (8 bits).
     *
     * @returns The current 8-bit links value (0-255).
     */
    async getLinks(): Promise<number> {
        const response = await promisify<Record<string, never>, LinksState>(
            this._stub as unknown as Record<string, Function>,
            "getLinks",
            {},
        );
        return response.value;
    }

    /**
     * Set the raw keyboard links byte (8 bits).
     *
     * @param value - The 8-bit links value (0-255).
     * @returns True if successful.
     * @throws Error if value is not in range 0-255 or server rejects.
     */
    async setLinks(value: number): Promise<boolean> {
        if (value < 0 || value > 255) {
            throw new Error("Links value must be 0-255");
        }
        const response = await promisify<{ value: number }, SetLinksResponse>(
            this._stub as unknown as Record<string, Function>,
            "setLinks",
            { value },
        );
        if (!response.success) {
            throw new Error(response.error);
        }
        return true;
    }

    /**
     * Get the startup screen mode (0-7).
     *
     * @returns The configured startup screen mode.
     */
    async getStartupScreenMode(): Promise<number> {
        const response = await promisify<Record<string, never>, StartupScreenModeState>(
            this._stub as unknown as Record<string, Function>,
            "getStartupScreenMode",
            {},
        );
        return response.mode;
    }

    /**
     * Set the startup screen mode (0-7).
     *
     * @param mode - The screen mode (0-7).
     * @returns True if successful.
     * @throws Error if mode is not in range 0-7 or server rejects.
     */
    async setStartupScreenMode(mode: number): Promise<boolean> {
        if (mode < 0 || mode > 7) {
            throw new Error("Mode must be 0-7");
        }
        const response = await promisify<{ mode: number }, SetStartupScreenModeResponse>(
            this._stub as unknown as Record<string, Function>,
            "setStartupScreenMode",
            { mode },
        );
        if (!response.success) {
            throw new Error(response.error);
        }
        return true;
    }

    /**
     * Check if auto-boot on SHIFT-BREAK is enabled.
     *
     * @returns True if SHIFT-BREAK triggers auto-boot.
     */
    async getStartupAutoBoot(): Promise<boolean> {
        const response = await promisify<Record<string, never>, StartupAutoBootState>(
            this._stub as unknown as Record<string, Function>,
            "getStartupAutoBoot",
            {},
        );
        return response.enabled;
    }

    /**
     * Enable or disable auto-boot on SHIFT-BREAK.
     *
     * When enabled, SHIFT-BREAK causes MOS to load and run !Boot.
     *
     * @param enabled - True to enable auto-boot.
     * @returns True if successful.
     */
    async setStartupAutoBoot(enabled: boolean): Promise<boolean> {
        const response = await promisify<{ enabled: boolean }, SetStartupAutoBootResponse>(
            this._stub as unknown as Record<string, Function>,
            "setStartupAutoBoot",
            { enabled },
        );
        return response.success;
    }
}
